Fix pixelwidth and pixelheight for raf files

astippich opened the pull request:
Extract the pixel width and pixel height of raf images from the raf metadata found in fuji raw files.
Code inspired by exiftool

Fixes #755

piponazo commented on src/rafimage.cpp:
```diff
@@ -296,6 +292,37 @@ namespace Exiv2 {
         long jpg_img_len = Exiv2::getULong((const byte *) jpg_img_length, bigEndian);
+        byte cfa_header_offset [4];
+        io_->read(cfa_header_offset, 4);
+        // byte cfa_header_length [4];
```
Please remove the commented lines

piponazo commented on src/rafimage.cpp:
```diff
@@ -296,6 +292,37 @@ namespace Exiv2 {
+        byte byte_tag[2];
+        byte byte_size[2];
+        int32_t count = getLong(byte_count, bigEndian);
+        for (int32_t i = 0; i < count; ++i ) {
```
I wonder if this loop here could affect to the call `TiffParser::decode` we have later in line 307 :thinking:

D4N replied:
Unlikely, as `TiffParser::decode` shares no internal state with `this` and all the parameters appear to be untouched by the new code.

D4N replied:
Unlikely, as `TiffParser::decode` shares no internal state with `this` and all the parameters appear to be untouched by the new code.

piponazo reviewed (changes requested):
We would need to have a new system test (check tests under the `tests` folder) to make sure that your changes are fixing the issue. 

Normally we create first a python test for reproducing the detected issue and then you can apply the fix (preferentially the fix and the test in different commits). In #792 you can find an example of this procedure. Note that for providing a sample image, we remove the image content and we just keep the metadata.

D4N commented on src/rafimage.cpp:
```diff
@@ -296,6 +292,37 @@ namespace Exiv2 {
         long jpg_img_off = Exiv2::getULong((const byte *) jpg_img_offset, bigEndian);
         long jpg_img_len = Exiv2::getULong((const byte *) jpg_img_length, bigEndian);
+        byte cfa_header_offset [4];
+        io_->read(cfa_header_offset, 4);
```
Before doing all these reads, please check that the image is actually large enough, or check the return value of each `io_->read()`.

D4N commented on src/rafimage.cpp:
```diff
@@ -296,6 +292,37 @@ namespace Exiv2 {
+        io_->read(byte_count, 4);
+        byte byte_tag[2];
+        byte byte_size[2];
+        int32_t count = getLong(byte_count, bigEndian);
```
Please add a sanity check for this value, otherwise this will create a near-infinite loop.

astippich replied:
What is a sane value for you? The used test file has 32 tags, but I have no clue if other RAF files have a different number of raf tags.

D4N replied:
If there is no upper limit mandated by a standart, then you can make a rough check like: `enforce(count < remaining_file_size/size_of_a_tag);`, provided that this is at least an order of magnitude less than the maximum value of `int32_t` for all reasonable files.

If I understand your code, then aren't you searching for the pixelwidth and pixelheight inside the `cfa_hdr`? And you've just extracted its length a few lines above, so given the tag size and the length, that gives you a pretty good upper limit on the number of tags that can possibly exist in there.

astippich replied:
tag size is not constant according to exiftool, but I guess I could add a check with the minimal size for at least something

D4N commented on src/rafimage.cpp:
```diff
@@ -296,6 +292,37 @@ namespace Exiv2 {
+                pixelWidth_ = getUShort(image_width, bigEndian);
+                break;
+            } else {
+                io_->seek(tag_size, BasicIo::cur);
```
Please add a check that this seek works and doesn't put you outside of the image. Mind that `tag_size` is extracted from the image and thus potentially malicious. If it actually puts you outside of the image then all further reads will return garbage values and potentially result in an infinite loop.

D4N reviewed (changes requested):
Thanks for your contribution!

I've added some comments myself. A test case would be nice as @piponazo noted.

astippich commented:
Thanks for the feedback. I will try to implement it, but I am quite busy the next weeks.

Regarding the unit test, I expected and somewhat feared this request :)
1. I implemented it for a user of KDE's KFileMetaData and tested it using the user's file (https://bugs.kde.org/show_bug.cgi?id=380919) which I would rather not like to upload as a test file, so I basically have no test file for this
2. Since these are raw file they are quite big. I do not know how to reduce the file size

Any suggestions welcome.

D4N commented:
> Thanks for the feedback. I will try to implement it, but I am quite busy the next weeks.
> 
> Regarding the unit test, I expected and somewhat feared this request :)

You don't have to write a unit test, an integration test would be easier imho. We have a custom integration test framework written in Python (take a look at the Python files in `tests/bugfixes/`). You essentially just feed it a `exiv2` invocation and the expected output and the framework does the rest for you.

> 
> 1. I implemented it for a user of KDE's KFileMetaData and tested it using the user's file (https://bugs.kde.org/show_bug.cgi?id=380919) which I would rather not like to upload as a test file, so I basically have no test file for this
> 
> 2. Since these are raw file they are quite big. I do not know how to reduce the file size

`exiv2 -ex $your_image` will extract the metadata from the image and drop all pixel contents, so unless they put their credit card number in the metadata, it should be fine to upload. You might want to check the files contents nevertheless.

> 
> 
> Any suggestions welcome.

astippich commented:
> `exiv2 -ex $your_image` will extract the metadata from the image and drop all pixel contents, so unless they put their credit card number in the metadata, it should be fine to upload. You might want to check the files contents nevertheless.

That does not seem to work, probably because it is a raw file and the size information is also in the custom RAF metadata. I have a RAF image which the user granted me the rights to upload, but it is 31 MB. Is this a problem?

D4N commented:
> I have a RAF image which the
> user granted me the rights to upload, but it is 31 MB. Is this a problem?

Kind of, we do not want to add a 31 MB file to git and even for git LFS it's
quite a huge file (I wouldn't want too many files larger than 1 MB, so that we
won't hit the 2GB limit anytime soon).

Is it possible to scale this file down, e.g. to 100x100 px?

astippich commented:
> > I have a RAF image which the
> > user granted me the rights to upload, but it is 31 MB. Is this a problem?
> 
> Kind of, we do not want to add a 31 MB file to git and even for git LFS it's
> quite a huge file (I wouldn't want too many files larger than 1 MB, so that we
> won't hit the 2GB limit anytime soon).
> 
> Is it possible to scale this file down, e.g. to 100x100 px?

According to the user, no (I asked for lowest quality and resolution). Again, probably because it is a raw file. I found https://rawsamples.ch/index.php/en/fuji which is content provided under Creative Commons and provides a 6 MB raf file. Is this acceptable?

D4N commented:
> According to the user, no (I asked for lowest quality and resolution). Again,
> probably because it is a raw file. I found
> https://rawsamples.ch/index.php/en/fuji which is content provided under Creative
> Commons and provides a 6 MB raf file. Is this acceptable?

Unfortunately 6 MB is still too much to be checked into git.

I've created #896 to enable git lfs, so that we can add your test file.

astippich committed:
read image width and height from RAF metadata

astippich committed:
fix test for issue 857

astippich commented:
Finally found some time again to work on this. I've rebased and added error handling as requested.
As a consequence, the test for issue 857 failed differently, and I've adjusted the test.

piponazo commented:
Thanks @astippich for the contribution. The changes look good to me (the formatting is not satisfying the clang-format style we have defined, but that's something we could clean-up later, unless you do not mind to add an additional commit applying clang-format to your additions [not to the complete file]).

However as @D4N already mentioned, we should not include such a big binary file in the repository. Until we add support for git LFS, or other strategy to deal with such big files, I would propose to drop the commit including the test and the binary file. Maybe we could keep a patch of that commit in this PR for future reference. Or if you guys have any other better idea, it will be welcomed.

D4N commented on tests/bugfixes/github/test_issue_755.py:
```diff
@@ -0,0 +1,37 @@
+class RAFPixelWidthAndHeight(metaclass=system_tests.CaseMeta):
+    url = "https://github.com/Exiv2/exiv2/issues/755"
+
+    commands = ["$exiv2 ../../../test/data/issue_755.RAF"]
```
```suggestion
    commands = ["$exiv2 $data_path/issue_755.RAF"]
```

D4N commented on tests/bugfixes/github/test_issue_755.py:
```diff
@@ -0,0 +1,37 @@
+
+    commands = ["$exiv2 ../../../test/data/issue_755.RAF"]
+    stdout = [
+        """File name       : ../../../test/data/issue_755.RAF
```
```suggestion
        """File name       : $data_path/issue_755.RAF
```

D4N commented on src/rafimage.cpp:
```diff
@@ -313,6 +309,52 @@ namespace Exiv2 {
+
+        long cfa_hdr_off = static_cast<long>(cfa_hdr_off_u32);
+
+        if (io_->seek(cfa_hdr_off,BasicIo::beg) != 0)  throw Error(kerFailedToReadImageData);;
```
```suggestion
        if (io_->seek(cfa_hdr_off,BasicIo::beg) != 0)  throw Error(kerFailedToReadImageData);
```

D4N reviewed (changes requested):
This looks better now, thanks for the changes.

- Please add an explanation to the chunk `readMetadata()` what you are doing there (something high level and if it makes sense, link to the issue that you reported).
- Consider using the `--grep ` function from `exiv2` in the test file to grep for the relevant metadata only.

D4N commented:
> However as @D4N already mentioned, we should not include such a big binary file in the repository. Until we add support for git LFS, or other strategy to deal with such big files, I would propose to drop the commit including the test and the binary file. Maybe we could keep a patch of that commit in this PR for future reference. Or if you guys have any other better idea, it will be welcomed.

Not really, beside waiting until we enable git LFS.

astippich commented:
> This looks better now, thanks for the changes.
> 
>     * Please add an explanation to the chunk `readMetadata()` what you are doing there (something high level and if it makes sense, link to the issue that you reported).
> 
>     * Consider using the `--grep ` function from `exiv2` in the test file to grep for the relevant metadata only.

D4N commented:
> I added a sentence to the code, if you want more let me know
> 

I was hoping for a something that will explain the future maintainer (i.e. me or someone else from the team) how the format looks like that you are parsing. Currently, I have to read all your code and try to find out what it does. If you could explain that for someone that doesn't have your knowledge, that would be greatly appreciated.

> I tried that, but could not find any documentation on what tag name to use. Could it be that this also works only for Exif tag names?

@clanmills do you know that?

> 
> I will open a separate PR for the test so that you can merge when the infrastructure is ready.

I think setting up git lfs has been long overdue, I'll try to look into that soon.

astippich committed:
document raf metadata structure

astippich commented:
Anything to do here from my side?

piponazo commented:
Sorry for the delay on getting this merged @astippich . I'll let @D4N take care of merging it, since the PR is pending on his approval.

D4N reviewed (approved):
Thanks for your patience and your effort!
